File: Ruby/Hangar.rb
#encoding: UTF-8

# Autores: Manuel Díaz-Meco Terrés y Ana Graciani Donaire
# 2ºDGIIM		PDOO

# Esta clase representa el Hangar y sus múltiples atributos así como
# el espacio del que se dispone y las armas que se tienen

require_relative 'HangarToUI'

module Deepspace
	class Hangar
	
		# Constructor
		
		def initialize (capacity)
			@maxElements = capacity
			# Atributos de referencia
			@weapons = Array.new # otra opción: @weapons = []
			@shieldBoosters = Array.new
		end	
		
		# Método de copia (método de clase)
		def self.newCopy(h)
			copy = Hangar.new(h.maxElements)
			h.weapons.each {|x| copy.addWeapon(x)}
			h.shieldBoosters.each {|x| copy.addShieldBooster(x)}
			return copy
		end
		
		private
		
		# Devuelve true si aún hay esacio para añadir elementos y que
		# por lo tanto no se ha llegado a la capacidad máxima
		def spaceAvailable
			return (@weapons.size + @shieldBoosters.size) < @maxElements
		end
		
		public
		
		# Añade el arma al hangar si queda espacio en el Hangar, devolviendo
		# true en ese caso. Devuelve false en cualquier otro caso
		
		def addWeapon(w)
			rvalue = false
			if spaceAvailable
				@weapons.push(w) # devuelve el array
				rvalue = true
			end
			return rvalue
		end
		
		# Añade el potenciador de escudo al hangar si queda espacio. 
		# Devuelve true si ha sido posible añadir el potenciador, false en
		# otro caso
		
		def addShieldBooster(s)
			rvalue = false
			if spaceAvailable
				@shieldBoosters.push(s)
				rvalue = true
			end
			return rvalue
		end
		
		# Consultores
		
		def maxElements
			@maxElements
		end
		
		def shieldBoosters
			@shieldBoosters
		end
		
		def weapons
			@weapons
		end
		
		# Elimina el potenciador de escudo número s del hangar y lo devuelve, 
		# siempre que este exista. Si el ínidice suminstrado es incorrecto
		# devuelve nil
		
		def removeShieldBooster(s)
			return @shieldBoosters.delete_at(s)
		end
		
		# Elimina el arma número w del hangar y la devuelve, siempre que esta
		# exista. Si el índice suminstrado es incorrecto devuelve nil
		
		def removeWeapon(w)
			return @weapons.delete_at(w)
		end

		def getUIversion
			return HangarToUI.new(self)
		end

		def to_s
			return getUIversion.to_s
		end
		
	end

	
	# código de prueba (constructor de copia)
=begin
	require_relative 'WeaponType'
	require_relative 'Weapon'
	require_relative 'ShieldBooster'

	hangar1 = Hangar.new(3)
	puts "Hangar 1 vacio"
	puts hangar1.inspect
	puts hangar1.to_s
	wt = WeaponType::LASER
	weapon = Weapon.new("meinol", wt, 3)
	shieldb = ShieldBooster.new("sunage", 2.0, 3)
	hangar1.addWeapon(weapon)
	hangar1.addShieldBooster(shieldb)
	hangar2 = Hangar.newCopy(hangar1)

	puts "Hangar 1"
	puts hangar1.inspect
	puts hangar1.to_s
	puts "Hangar 2"
	puts hangar2.inspect
	puts hangar2.to_s
=end

end

File: Java/deepspace/src/deepspace/Hangar.java
/* @authors Manuel Diaz-Meco Terrés y Ana Graciani Donaire
 * 2ºDGIIM         PDOO
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */

package deepspace;

import java.util.ArrayList;

/**
 * Esta clase representa el Hangar y sus múltiples atributos así como el espacio
 * del que se dispone y las armas que se tienen
 * @authors Manuel Diaz-Meco Terrés y Ana Graciani Donaire
 */

public class Hangar {
    
    private int maxElements;
    
    // Atributos de referencia
    
    private ArrayList<Weapon> weapons = new ArrayList<>();
    private ArrayList<ShieldBooster> shieldBoosters = new ArrayList<>();
    
    /**
     * Constructor con parámetros
     * @param capacity número máximo de elementos del hangar
     */
    Hangar(int capacity) { maxElements = capacity; }
    
    /**
     * Constructor de copia
     * @param h Hangar a copiar
     */
    Hangar(Hangar h) {
        maxElements = h.maxElements;
        weapons = h.getWeapons();
        shieldBoosters = h.getShieldBoosters();
    }
    
    HangarToUI getUIversion() {
        return (new HangarToUI(this));
    }    
    
    /**
     * Indica si hay espacio disponible en el Hangar
     * @return Devuelve true si el número de elementos es menor que maxElements
     * y false si es igual
     */   
    private boolean spaceAvailable() {
        return (weapons.size()+shieldBoosters.size() < maxElements);
    }

      
    /**
     * Añade el arma al Hangar si queda espacio
     * @param w Arma a añadir
     * @return true (se ha añadido) / false (no se ha podido añadir)
     */    
    public boolean addWeapon(Weapon w) {
        if (spaceAvailable())
            return weapons.add(w);
       
        return false;
    }
    
    /**
     * Añade el potenciador de escudo al Hangar si queda espacio
     * @param s Potenciador de escudo a añadir
     * @return true (se ha añadido) / false (no se ha podido añadir)
     */
    public boolean addShieldBooster(ShieldBooster s) {
        if (spaceAvailable())
            return shieldBoosters.add(s);
        
        return false;
    }
    
    /**
     * Consultor de maxElements (capacidad del hangar)
     * @return maxElements
     */
    public int getMaxElements() { return maxElements; }   
    
    /**
     * Consultor de shieldBoosters (colección de potenciadores de escudo)
     * @return shieldBoosters
     */    
    public ArrayList<ShieldBooster> getShieldBoosters() {
        return shieldBoosters;
    }
        
    /**
     * Consultor de weapons (colección de armas)
     * @return weapons
     */    
    public ArrayList<Weapon> getWeapons() {
        return weapons;
    }
        
    /**
     * Elimina el potenciador de escudo número s del hangar y lo devuelve,
     * siempre que este exista, si el índice suministrado es incorrecto
     * devuelve null.
     * @param s índice del shieldBooster a eliminar
     * @return elemento eliminado / null si no se pudo eliminar
     */    
    public ShieldBooster removeShieldBooster(int s) {
        if (0 <= s && s < shieldBoosters.size()) {
            return shieldBoosters.remove(s); // devuelve el obj, eliminado
        }
        return null; 
    }
    
    /**
     * Elimina el arma número w del hangar y la devuelve, siempre que esta
     * exista, si el índice suministrado es incorrecto devuelve null.
     * @param w índice del arma a eliminar
     * @return elemento eliminado / null si no se pudo eliminar
     */ 
    public Weapon removeWeapon(int w) {
        if (0 <= w && w < weapons.size()) {
            return weapons.remove(w);
        }
        return null; 
    } 
    
    @Override
    public String toString(){
        String result;
        
        result =  "maxElements: " + maxElements;
        if (weapons != null)
                result += ", weapons:" + weapons.toString();
        if (shieldBoosters != null)
                result +=  ", shieldBoosters: " + shieldBoosters.toString();
        return result;
    }
}